examples/linear-maps.dx

```diff
@@ -0,0 +1,391 @@
+import linalg
+
+'## Instances for triangular tables
+How to avoid these repetitive instances?
+
+instance {a n} [Add a] Add (i:n => (..<i) => a)
+  add = \xs ys. for i. xs.i + ys.i
+  sub = \xs ys. for i. xs.i - ys.i
+  zero = for _. zero
+
+instance {a n} [Add a] Add (i:n => (i<..) => a)
+  add = \xs ys. for i. xs.i + ys.i
+  sub = \xs ys. for i. xs.i - ys.i
+  zero = for _. zero
+
+instance {a n} [VSpace a] VSpace (i:n => (..i) => a)
+  scale_vec = \s xs. for i. s .* xs.i
+
+instance {a n} [VSpace a] VSpace (i:n => (i..) => a)
+  scale_vec = \s xs. for i. s .* xs.i
+
+instance {a n} [VSpace a] VSpace (i:n => (..<i) => a)
+  scale_vec = \s xs. for i. s .* xs.i
+
+instance {a n} [VSpace a] VSpace (i:n => (i<..) => a)
+  scale_vec = \s xs. for i. s .* xs.i
+
+instance {n a} [Arbitrary a] Arbitrary (i:n=>(..<i) => a)
+  arb = \x. for i. arb $ new_key (ordinal i)
+
+instance {n a} [Arbitrary a] Arbitrary (i:n=>(..i) => a)
+  arb = \x. for i. arb $ new_key (ordinal i)
+
+instance {n a} [Arbitrary a] Arbitrary (i:n=>(i..) => a)
+  arb = \x. for i. arb $ new_key (ordinal i)
+
+instance {n a} [Arbitrary a] Arbitrary (i:n=>(i<..) => a)
+  arb = \x. for i. arb $ new_key (ordinal i)
+
+
+t = (i:(Fin 2)=> (..i) => Float32)
+x : t = arb $ new_key 0
+
+def f (g:t) : t =
+  g
+
+f x
+
+
+'Todo: factor out a more general linear map typeclass, and make this one inherit from it.
+-- interface [VSpace m, VSpace in, VSpace out] LinearMap m in out
+  -- apply: m -> in -> out
+  -- transpose': m -> m
+
+'## Linear Endomorphisms
+a.k.a. linear maps from a space back to that same space.
+
+
+interface [VSpace m, VSpace v] LinearEndo m v
+  apply: m -> v -> v
+  -- determinant': m -> Float
+  -- transposeType: Type
+  -- transpose': m -> transposeType
+  -- identity': m
+  diag: m -> v
+  solve': m -> v -> v
+
+'We'd like to remove `v` from the above interface,
+and instead use associated types to specify a `v` for each `m`.
+But for now, fields in typeclasses can't refer to one another.
+This means that `determinant'` can't be part of this typeclass yet,
+because `v` is always ambiguous at its usage site.
+
+
+instance {v} [Mul v, VSpace v] LinearEndo (Float) v
+  apply = (.*)
+  diag = \a. a .* one
+  solve' = \a b. b / a
+
+-- Diagonal matrices
+instance {n v} [Mul v, VSpace v] LinearEndo (n=>Float) (n=>v)
```

Comment:
I'm not sure how I feel with the assumption that all 1D vectors represent diagonal matrices. Perhaps make a newtype?